lib/codecs/url.js

```diff
@@ -0,0 +1,149 @@
+
+// URL prefix encode/decode map
+urlHeaders.set('http://', 1);
+urlHeaders.set(1, 'http://');
+urlHeaders.set('https://', 2);
+urlHeaders.set(2, 'https://');
+urlHeaders.set('did:v1:nym', 1024);
+urlHeaders.set(1024, 'did:v1:nym');
+urlHeaders.set('did:key', 1025);
+urlHeaders.set(1025, 'did:key');
+
+// Base58 DID URL encoding/decoding class
+const b58urlRegex =
+  /(did:.*)?:z([1-9A-HJ-NP-Za-km-z]+)?#{0,1}z{0,1}([1-9A-HJ-NP-Za-km-z]+)?/g;
+export class Base58DidUrlCodec {
+  encode(value) {
+    // break the base58 DID URL into segments
+    const matches = Array.from(value.matchAll(b58urlRegex));
+    const match = matches[0];
+
+    // encode the DID URL segments
+    const encodedUrl = (urlHeaders.has(match[1])) ? urlHeaders.get(match[1]) :
+      match[1];
+    const encodedDid = b58.decode(match[2]);
+    const encodedFragment = (match[3]) ? b58.decode(match[3]) : undefined;
+
+    // construct the encoded URL array
+    const encodedValue = [encodedUrl, encodedDid];
+    if(encodedFragment) {
+      encodedValue.push(encodedFragment);
+    }
+    return encodedValue;
+  }
+
+  decode(value) {
+    const [encodedHeader, encodedAuthority, encodedFragment] = value;
+
+    const header = urlHeaders.get(encodedHeader);
+    const authority = b58.encode(Uint8Array.from(encodedAuthority));
+    const fragment = (encodedFragment) ?
+      b58.encode(Uint8Array.from(encodedFragment)) : undefined;
+    const decodedValue = (fragment) ?
+      `${header}:${authority}#${fragment}` :
+      `${header}:${authority}`;
+
+    return decodedValue;
+  }
+}
+
+// URL prefix detection to map prefix to encoding/decoding class
+// FIXME: urlEncoders and urlDecoders need to be refactored
+urlEncoders.push([/did:v1:nym:z/, Base58DidUrlCodec]);
+urlEncoders.push([/did:key:z/, Base58DidUrlCodec]);
+
+urlDecoders.push([1024, Base58DidUrlCodec]);
+urlDecoders.push([1025, Base58DidUrlCodec]);
+
+// encodes a URL given a term coding map
+function _encodeUrl(url, termCodecMap) {
+  // determine if a term value exists in the term coding map
+  if(termCodecMap.get(url)) {
+    return termCodecMap.get(url).value;
+  }
+
+  // attempt to use more general URL encoders
+  let encodedValue = url;
+  const encoders = urlEncoders.filter(element => url.match(element[0]));
+  if(encoders.length === 1) {
+    const codec = new encoders[0][1];
+    encodedValue = codec.encode(url);
+  }
+
+  return encodedValue;
+}
+
+// decode a URL given a term codec map
+function _decodeUrl(encodedUrl, termCodecMap) {
+  // determine if a value exists in the term coding map
+  if(termCodecMap.get(encodedUrl)) {
+    return termCodecMap.get(encodedUrl).value;
+  }
+
+  // attempt to use more general URL encoders
+  let decodedValue = encodedUrl;
+  const decoders = urlDecoders.filter(element => encodedUrl[0] === element[0]);
+  if(decoders.length === 1) {
+    const codec = new decoders[0][1];
+    decodedValue = codec.decode(encodedUrl);
+  }
+
+  return decodedValue;
```

Comment:
Is it by designed that the `encodedUrl` is returned if there is no decoder?  Or is this an error condition?  Same applies to `_encodeUrl`

Reply:
The documentation was bad, fixed in ac36160ff9321e5c1d2656a0a3013ae5421c34d8. Basically, use a more aggressive encoder/decoder if one is found... otherwise, just pass the value through w/o encoding/decoding.